### PR TITLE
ARROW-13827: [C++][Gandiva] Implement LEVENSHTEIN Hive functions on Gandiva

### PR DESCRIPTION
**Pull request by jpedroantunes**

### CODE REVIEW

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 1]
```diff
@@ -1642,6 +1642,55 @@ const char* convert_toUTF8(int64_t context, const char* value, int32_t value_len
...
+      return 0;
+    }
+  }
+  for (int32_t i = 0; i <= in1_len; i++) dist[i][0] = i;
+  for (int32_t j = 1; j <= in2_len; j++) dist[0][j] = j;
+
+  for (int32_t j = 0; j < in2_len; j++) {
+    for (int32_t i = 0; i < in1_len; i++) {
+      if (in1[i] == in2[j]) {
+        dist[i + 1][j + 1] = dist[i][j];
```
Comment (rkavanap) [inline cpp/src/gandiva/precompiled/string_ops.cc:1674]: shouldn't this also be min(dist[i][j], 1 + dist[i][j + 1], 1 + dist[i + 1][j]), when in1[i] = in2[j]?
and when in1[i] != in2[j]  it is a min (1 + dist[i][j], 1 + dist[i][j + 1], 1 + dist[i + 1][j]) as properly coded from 1676..1680... 

Atleast In Java it is coded that way..

File: cpp/src/gandiva/precompiled/string_ops_test.cc [hunk 2]
```diff
@@ -912,6 +912,35 @@ TEST(TestStringOps, TestReverse) {
...
+  EXPECT_EQ(levenshtein_utf8_utf8(ctx_ptr, "test", 4, "task", 4), 2);
+  EXPECT_FALSE(ctx.has_error());
+
+  EXPECT_EQ(levenshtein_utf8_utf8(ctx_ptr, "cat", 3, "coat", 4), 1);
+  EXPECT_FALSE(ctx.has_error());
+
+  EXPECT_EQ(levenshtein_utf8_utf8(ctx_ptr, "coat", 4, "coat", 4), 0);
+  EXPECT_FALSE(ctx.has_error());
+
+  EXPECT_EQ(levenshtein_utf8_utf8(ctx_ptr, "book", -5, "back", 4), 0);
```
Comment (rkavanap) [inline cpp/src/gandiva/precompiled/string_ops_test.cc:937]: test case for "book" to "bo" or "boo". and "b" to "book"?

File: cpp/src/gandiva/tests/projector_test.cc [hunk 3]
```diff
@@ -818,6 +818,41 @@ TEST_F(TestProjector, TestConcat) {
...
+  auto lev_expr = TreeExprBuilder::MakeExpression("levenshtein", {f0, f1}, field_lev);
+
+  std::shared_ptr<Projector> projector;
+  auto status = Projector::Make(schema, {lev_expr}, TestConfiguration(), &projector);
+  EXPECT_TRUE(status.ok()) << status.message();
+
+  // Create a row-batch with some sample data
+  int num_records = 4;
+  auto array0 = MakeArrowArrayUtf8({"cat", "task", "", "a"}, {true, true, true, true});
+  auto array1 = MakeArrowArrayUtf8({"coat", "test", "a", ""}, {true, true, true, true});
```
Comment (rkavanap) [inline cpp/src/gandiva/tests/projector_test.cc:840]: maybe some fields with validity bit false as well (i.e null fields)?

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 4]
```diff
@@ -1642,6 +1642,55 @@ const char* convert_toUTF8(int64_t context, const char* value, int32_t value_len
   return value;
 }
 
+// Calculate the levenshtein distance between two string values
+FORCE_INLINE
+gdv_int32 levenshtein_utf8_utf8(int64_t context, const char* in1, int32_t in1_len,
```
Comment (rkavanap) [inline cpp/src/gandiva/precompiled/string_ops.cc:1647]: I feel there is a more space efficient algorithm than this as we are only interested in the edit distance and is not interested in reconstructing the edited input..Atleast the Java version uses this space efficient algorithm.

Is there any reason why we have to use this algorithm and not the one used by Java?

Comment (rkavanap) [reply cpp/src/gandiva/precompiled/string_ops.cc:1647]: Infact I feel more strongly now that we should use a memory efficient algorithm. What if the column width is 65K? we will need 4GB to run this alorithm. In fact Java version uses an algorithm that uses only 2n space where n is the smallest of the 2 strings.

**Comment by vvellanki** [comment]
How is this different from https://github.com/apache/arrow/pull/11522?

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 4 above)
Comment (rkavanap) [reply cpp/src/gandiva/precompiled/string_ops.cc:1647]: https://github.com/apache/commons-lang/blob/18a82dc4ae9c1095216b6cb288b3c5567ffd179b/src/main/java/org/apache/commons/lang3/StringUtils.java#L2336 has an implementation of levenshtein that use less memory.

**Pull request closed without merging**